crates/core/tedge_api/src/entity_store.rs

```diff
@@ -676,6 +672,164 @@ impl EntityStore {
     }
 }
 
+/// In-memory representation of the entity tree
+struct EntityTree {
+    entities: HashMap<EntityTopicId, EntityNode>,
+    entity_id_index: HashMap<EntityExternalId, EntityTopicId>,
+}
+
+#[derive(Debug)]
+struct EntityNode {
+    metadata: EntityMetadata,
+    children: HashSet<EntityTopicId>,
+}
+
+impl EntityNode {
+    pub fn new(metadata: EntityMetadata) -> Self {
+        EntityNode {
+            metadata,
+            children: HashSet::new(),
+        }
+    }
+
+    pub fn metadata(&self) -> &EntityMetadata {
+        &self.metadata
+    }
+    pub fn mut_metadata(&mut self) -> &mut EntityMetadata {
+        &mut self.metadata
+    }
+}
+
+impl EntityTree {
+    /// Create the tree of entities for the main device, given its name, topic id and metadata
+    pub fn new(
+        device_name: EntityExternalId,
+        topic_id: EntityTopicId,
+        metadata: EntityMetadata,
+    ) -> Self {
+        EntityTree {
+            entities: HashMap::from([(topic_id.clone(), EntityNode::new(metadata))]),
+            entity_id_index: HashMap::from([(device_name, topic_id)]),
+        }
+    }
+
+    pub fn contains_key(&self, topic_id: &EntityTopicId) -> bool {
+        self.entities.contains_key(topic_id)
+    }
+
+    /// Returns information about an entity under a given MQTT entity topic identifier.
+    pub fn get(&self, entity_topic_id: &EntityTopicId) -> Option<&EntityMetadata> {
+        self.entities.get(entity_topic_id).map(EntityNode::metadata)
+    }
+
+    /// Returns a mutable reference to the `EntityMetadata` for the given `EntityTopicId`.
+    fn get_mut(&mut self, entity_topic_id: &EntityTopicId) -> Option<&mut EntityMetadata> {
+        self.entities
+            .get_mut(entity_topic_id)
+            .map(EntityNode::mut_metadata)
+    }
+
+    /// All the entities with a given parent.
+    pub fn children(&self, parent_id: &EntityTopicId) -> Vec<(&EntityTopicId, &EntityMetadata)> {
+        let Some(children) = self.entities.get(parent_id).map(|node| &node.children) else {
+            return vec![];
+        };
+
+        children
+            .iter()
+            .filter_map(|topic_id| self.entities.get_key_value(topic_id))
+            .map(|(k, v)| (k, v.metadata()))
+            .collect()
+    }
+
+    /// Returns information for an entity under a given device/service id.
+    pub fn get_by_external_id(&self, external_id: &EntityExternalId) -> Option<&EntityMetadata> {
+        let topic_id = self.entity_id_index.get(external_id)?;
+        self.get(topic_id)
+    }
+
+    /// Insert a new entity
+    ///
+    /// Return Inserted if the entity is new
+    /// Return Updated if the entity was previously registered and has been updated by this call
+    /// Return Unchanged if the entity not affected by this call
+    pub fn insert(
+        &mut self,
+        external_id: EntityExternalId,
```

Comment:
Is this device name / external id required?

I kept it to reduce the amount of changes, but this external id is also given by the metadata.